### DISCUSSION TITLE
Is there a built-in option to have AutoFixture randomly supply null instead of a value?

### DISCUSSION THREAD

**Discussion by macfma01** [AutoFixture/AutoFixture#1412] [Q&A] [answered]
It's realistic in some tests for some of the object properties to be null. AutoFixture seems to always supply a value for each type. Can it be set to alternate between values and nulls so that null cases are covered in the tests.

**Comment by aivascu** [comment, marked as answer]
@macfma01 no there isn't such an option however I think you might be interested in the `AutoFixture.Idioms` package.
The idioms encapsulate entire tests that you'd otherwise implement manually. One of these examples happens to be guard clause assertions (null checks).

Here's an example:
```cs
public class SomeEntity
{
    public SomeEntity(Guid uuid, string value)
    {
        Uuid = uuid;
        Value = value;
    }

    public Guid Uuid { get; }
    public string Value { get; private set; }

    public void DoStuff(string value)
    {
        this.Value = value;
    }
}
```
The test below will attempt to provide the `default` value to each arguments of each public constructor.
```cs
[Theory]
[AutoData]
public void GuardsConstructorArguments(GuardClauseAssertion assertion)
{
    assertion.Verify(typeof(SomeEntity).GetConstructors());
}
```
It could even test your entire type, including constructor, methods, and properties.
```cs
[Theory]
[AutoData]
public void GuardsConstructorArguments(GuardClauseAssertion assertion)
{
    assertion.Verify(typeof(SomeEntity).GetConstructors());
}
```
Keep in mind that the idioms package is highly opinionated in the implementation of it's default assertions, but you could use it to implement your own (less strict assertions).
You can find out more [here](https://docs.educationsmediagroup.com/unit-testing-csharp/autofixture/idioms).

**Comment by macfma01** [comment]
I thought I was on to something with this.

```fsharp
[<Fact>]
member this.``AutoFixture test`` () =
    
    fixture.Register<string>(fun x ->
        let rnd = System.Random()
        if rnd.Next(0, 100) < 50 then
            fixture.Create<string>()
        else
            null)
            
    let result = fixture.Create<string>()
    output.WriteLine($"result: {result}")
    test <@ result = null || result <> null @>
```
But it doesn't like me calling `fixture.Create<string>` from inside `fixture.Register<string>()`

`AutoFixture.ObjectCreationExceptionWithPath
AutoFixture was unable to create an instance of type AutoFixture.Kernel.SeededRequest because the traversed object graph contains a circular reference. Information about the circular path follows below. This is the correct behavior when a Fixture is equipped with a ThrowingRecursionBehavior, which is the default. This ensures that you are being made aware of circular references in your code. Your first reaction should be to redesign your API in order to get rid of all circular references. However, if this is not possible (most likely because parts or all of the API is delivered by a third party), you can replace this default behavior with a different behavior: on the Fixture instance, remove the ThrowingRecursionBehavior from Fixture.Behaviors, and instead add an instance of OmitOnRecursionBehavior:`

**Reply by aivascu** [reply]
@macfma01 I'm not an F# expert by any means but I think you could try the following.

```F#
let myFunction x = x + "some val"

let getMethod (name : string) (assembly : Assembly) =
    assembly.GetTypes()
    |> Seq.filter (fun t -> t.GetMethod(name) <> null)
    |> Seq.head
    |> fun t -> t.GetMethod(name)

[<Theory>]
[<AutoData>]
let ``My test`` (assertion : GuardClauseAssertion) =
    let assembly = Assembly.GetExecutingAssembly()
    let method = getMethod "myFunction" assembly
    assertion.Verify(method)
```
Result:
```
Message: 
AutoFixture.Idioms.GuardClauseException : An attempt was made to assign the value null to the parameter "x" of the method "myFunction", and no Guard Clause prevented this. Are you missing a Guard Clause?
Method Signature: System.String myFunction(System.String)
Parameter Type: System.String, System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e
Declaring Type: Tests, Acme.FuncTests, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null
Reflected Type: Tests, Acme.FuncTests, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null
```
